pennylane/templates/subroutines/qsvt.py

```diff
@@ -21,131 +21,203 @@
+             [ 0.     -0.975   0.      0.    ]
+             [ 0.      0.     -0.7158  0.    ]
+             [ 0.     -0.      0.     -0.975 ]]
+
+
+        Alternatively, if the input ``A`` is a matrix, the valid values for ``block_encoding`` are
+        ``"embedding"`` and ``"fable"``. In this case, the ``encoding_wires`` parameter corresponds to
+        the ``wires`` attribute in the templates :class:`~pennylane.BlockEncode` and :class:`~pennylane.FABLE`, respectively.
+        Note that for QSVT to work, the input matrix must be Hermitian.
+
+        .. code-block:: python
+
+            # P(x) = -1 + 0.2 x^2 + 0.5 x^4
+            poly = np.array([-0.1, 0, 0.2, 0, 0.5])
+
+            A = np.array([[-0.1, 0, 0, 0.1], [0, 0.2, 0, 0], [0, 0, -0.2, -0.2], [0.1, 0, -0.2, -0.1]])
+
+            dev = qml.device("default.qubit")
+
+            @qml.qnode(dev)
+            def circuit():
+                qml.qsvt(A, poly, encoding_wires=[0, 1, 2, 3, 4], block_encoding="fable")
+                return qml.state()
+
+            matrix = qml.matrix(circuit, wire_order=[0, 1, 2, 3, 4])()
+
+        .. code-block:: pycon
+
+            >>> print(np.round(matrix[:4, :4], 4).real)
+            [[-0.0954  0.     -0.0056 -0.0054]
+             [ 0.     -0.0912 -0.     -0.    ]
+             [-0.0056  0.     -0.0788  0.0164]
+             [-0.0054 -0.      0.0164 -0.0842]]
+    """
+
+    angles = qml.math.poly_to_angles(poly, "QSVT")
     projectors = []
 
-    if convention == "Wx":
-        angles = _qsp_to_qsvt(angles)
-        global_phase = (len(angles) - 1) % 4
+    # If the input A is a Hamiltonian
+    if hasattr(A, "pauli_rep"):
 
-        if global_phase:
-            with qml.QueuingManager.stop_recording():
-                global_phase_op = qml.GlobalPhase(-0.5 * np.pi * (4 - global_phase), wires=wires)
+        if block_encoding not in ["prepselprep", "qubitization", None]:
+            raise ValueError("block_encoding should take the value 'prepselprep' or 'qubitization'")
 
-    for idx, phi in enumerate(angles):
-        dim = c if idx % 2 else r
-        with qml.QueuingManager.stop_recording():
-            projectors.append(PCPhase(phi, dim=dim, wires=wires))
+        if block_encoding == "qubitization":
+            encoding = qml.Qubitization(A, control=encoding_wires)
 
-    projectors = projectors[::-1]  # reverse order to match equation
+        else:
+            encoding = qml.PrepSelPrep(A, control=encoding_wires)
 
-    if convention == "Wx" and global_phase:
-        return qml.prod(global_phase_op, QSVT(UA, projectors))
-    return QSVT(UA, projectors)
+        projectors = [
+            qml.PCPhase(angles[i], dim=2 ** len(A.wires), wires=encoding_wires + A.wires)
+            for i in range(len(angles))
+        ]
+
+    else:
+
+        if block_encoding not in ["embedding", "fable", None]:
+            raise ValueError("block_encoding should take the value 'embedding' or 'fable'")
+
```

Comment:
answered in previous comment:

"""
By definition, one is methods for block encoding arrays and the other for block encoding hamiltonians. I think that to avoid confusion, if a user wants to mix things up, they can do it outside the function
"""

Another problem that comes to mind is that we cannot predict the number of auxiliary qubit we will need since we do not know in advance the number of pauli terms we will have given a matrix

Reply:
I think it would be more natural to raise a not enough work wires error than to disallow this. But as mentioned above it would also be fine to specify how one could get around this in the error message.

Reply:
specified in the error message 👍